### PR TITLE
Reusing primitives for forward Batch Norm operator

### PR DESCRIPTION
**Pull request by kbinias**
Based on PR #12073

### CODE REVIEW

**Comment by luotao1** [comment]
Is #12515 the same as #12073 ？

**Comment by kbinias** [comment]
Yes.

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 1]
```diff
@@ -37,6 +37,122 @@ struct bn_type_traits {
...
+          *scaleshift_memory, *dst_memory);
+
+      dev_ctx_.SetBlob(prim_key, batch_norm_p);
+    } else {
+      is_reusing_ = true;
+    }
+    return batch_norm_p;
+  }
+
+  std::shared_ptr<batch_norm_fwd> AcquireTrainingBatchNormFwd(
```
Comment (mozga-intel) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc]: Does the function above is the similar. Maybe we are able to make one function, and only pass other parameters.

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 2]
```diff
@@ -37,6 +37,122 @@ struct bn_type_traits {
...
+          *batch_norm_pd_, *src_memory, *scaleshift_memory, *dst_memory,
+          *mean_memory, *variance_memory);
+
+      dev_ctx_.SetBlob(prim_key, batch_norm_p);
+    } else {
+      is_reusing_ = true;
+    }
+    return batch_norm_p;
+  }
+  //
```
Comment (mozga-intel) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc]: Remove this line

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 3]
```diff
@@ -37,6 +37,122 @@ struct bn_type_traits {
...
+    return dims2str(input_dims) + std::to_string(epsilon) +
+           std::to_string(flag) + std::to_string(is_test) +
+           std::to_string(format) + suffix;
+  }
+
+ private:
+  std::shared_ptr<batch_norm_fwd::primitive_desc> batch_norm_pd_;
+};
+
+std::string gethash(const memory::dims &input_dims, float epsilon,
```
Comment (mozga-intel) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc]: The name of this function is the same, the diff between them is visible inside, but those function has been doing the similar things - imo the name should be coherent. Change it from lowercase to uppercase, overload

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 4]
```diff
@@ -110,6 +217,14 @@ class BatchNormMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
     PADDLE_ENFORCE(scale_tz.size() == 1, "Dims of scale tensor is NOT 1");
     const unsigned int ic = scale_tz[0];
 
+    // MKLDNN requires a single piece of memory for scale and shift/bias data
+    const size_t scaleshift_size = 2 * ic;
```
Comment (mozga-intel) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc:194]: Can you tell me why we used 2?

Comment (kbinias) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc:194]: We have to preallocate one continuous piece of memory for 2 vectors, next we copy these vectors to scaleshift_data vector.

**Review by mozga-intel — APPROVED** [review]

**Comment by mozga-intel** [comment]
LGTM!
